tools/analyzer_plugin/test/integration/test_bases/server_plugin_contributor_test_base.dart

```diff
@@ -0,0 +1,79 @@
+import 'package:analyzer/src/generated/source.dart';
+import 'package:analyzer/src/generated/source_io.dart';
+import 'package:analyzer_plugin/protocol/protocol_common.dart';
+import 'package:analyzer_plugin/protocol/protocol_generated.dart';
+import 'package:meta/meta.dart';
+import 'package:over_react_analyzer_plugin/src/async_plugin_apis/assist.dart';
+import 'package:over_react_analyzer_plugin/src/diagnostic_contributor.dart';
+import 'package:path/path.dart' as p;
+import 'package:test/test.dart';
+
+import '../mocks.dart';
+import 'analysis_driver_test_base.dart';
+
+abstract class ServerPluginContributorTestBase extends AnalysisDriverTestBase {
+  PluginForTest get testPlugin => _plugin;
+
+  Source applySourceChange(SourceChange change, Source expectedSource) {
+    final path = p.normalize(expectedSource.uri.toFilePath());
+    final applicableFileEdits = change.edits.where((edit) => edit.file == path);
+    expect(applicableFileEdits.map((edit) => edit.file), contains(path),
+        reason: 'Expected SourceChange to include edits for "$path".');
+    final updated = SourceEdit.applySequence(
+      resourceProvider.getFile(path).readAsStringSync(),
+      [for (final fileEdit in applicableFileEdits) ...fileEdit.edits],
+    );
+    final file = resourceProvider.updateFile(path, updated);
+    return file.createSource();
+  }
+
+  SourceSelection createSelection(Source source, String target, {int offset}) {
```

Comment:
Some simple doc comments on `applySourceChange`/`createSelection` would be helpful for someone trying to understand what's going on in tests

Reply:
Will do!